fix(client,internals,migrate,generator-helper): handle multibyte UTF-8 characters split across chunk boundaries in byline

onozaty opened the pull request:
## Summary

Fixes #27695 

This PR fixes a bug where multibyte UTF-8 characters (e.g., Japanese text, emojis) were corrupted when split across stream chunk boundaries in the `byline` module used for parsing generator output.

## Problem

The original implementation used `chunk.toString(encoding)` which does not handle incomplete multibyte sequences correctly. When a multibyte character is split across chunk boundaries, the incomplete bytes are replaced with the Unicode replacement character (�), causing corruption of DMMF output containing non-ASCII characters.

## Solution

- Replace `chunk.toString(encoding)` with Node.js `StringDecoder` which properly buffers incomplete multibyte sequences
- Use `decoder.write()` in `_transform` to handle partial sequences
- Use `decoder.end()` in `_flush` to finalize any remaining buffered bytes
- Add `node:` prefix to built-in module imports for consistency

## Changes

**Modified packages**: `generator-helper`, `client`, `internals`, `migrate`

**Implementation:**
- **Fixed** `_transform` method to use `StringDecoder.write()`
- **Fixed** `_flush` method to append remaining bytes from `StringDecoder.end()`
- **Added** Comprehensive test coverage for multibyte character handling

## Test Coverage

Added 5 tests per package (20 total) covering:
- ✅ Single-byte ASCII characters
- ✅ Multibyte characters (Japanese, emoji) in complete chunks
- ✅ Multibyte characters split across chunk boundaries
- ✅ Truncated incomplete UTF-8 sequences at stream end

All tests pass in affected packages (`generator-helper`, `client`, `internals`, `migrate`).

onozaty committed:
test(generator-helper): add byline multibyte character handling tests

Add tests to verify LineStream behavior with multibyte UTF-8 characters.
Tests confirm that single-byte and multibyte characters work correctly
within single chunks, and expose a bug where multibyte characters split
across chunk boundaries become corrupted.

Test structure:
- Single-byte characters: both patterns pass
- Multibyte characters: single chunk passes, split boundary fails

onozaty committed:
fix(generator-helper): handle multibyte characters in byline streams

Use StringDecoder to properly handle incomplete multibyte UTF-8 sequences
that are split across chunk boundaries in streaming data. Without this,
characters like emoji and CJK text would be corrupted when split between
chunks.

Changes:
- Add StringDecoder import and instance to LineStream
- Replace Buffer.toString() with decoder.write() for proper streaming
- Finalize decoder in _flush() to handle remaining buffered bytes
- Assume single encoding per stream (encoding doesn't change mid-stream)

Fixes the test case where Japanese character "あ" split across chunks
would become replacement characters (���) instead of the correct character.

onozaty committed:
test(generator-helper): improve byline multibyte tests and simplify flush logic

- Add test for truncated multibyte sequences at stream end (tests _flush method)
- Simplify _flush logic by removing unreachable else branch
- Add clarifying comment about split() behavior guarantees

onozaty committed:
refactor(generator-helper): use node: prefix for built-in module imports in byline

onozaty committed:
fix(client,internals,migrate): handle multibyte UTF-8 characters split across chunk boundaries in byline

- Add StringDecoder to properly handle incomplete multibyte sequences
- Use node: prefix for built-in module imports
- Fix _transform to use decoder.write() instead of toString()
- Fix _flush to append remaining buffered bytes from decoder.end()
- Add comprehensive tests for multibyte character handling

onozaty committed:
fix(client,internals,migrate,generator-helper): handle encoding changes in byline StringDecoder

- Add _decoderEncoding to track current decoder encoding
- Recreate StringDecoder when encoding changes between chunks
- Add test coverage for encoding change scenarios (ascii to utf8)
- Addresses Copilot review feedback on PR #28535

onozaty committed:
test(client,internals,migrate): use consistent test() function naming in byline tests

- Replace it() with test() for consistency across all test files
- Addresses Copilot review feedback on PR #28535

onozaty commented on packages/migrate/src/utils/byline.ts:
```diff
@@ -128,6 +139,17 @@ LineStream.prototype._pushBuffer = function (encoding, keep, done) {
+      // Append remaining bytes to last line in buffer
+      // Note: _lineBuffer always has at least one element (even if empty string)
+      // because split() always returns an array with at least one element
+      this._lineBuffer[this._lineBuffer.length - 1] += remaining
```
Thank you for the careful review! Let me explain why `_lineBuffer` is guaranteed to have at least one element when we reach the `remaining` append logic:

**Analysis:**

1. `_lineBuffer` is initialized as an empty array: `this._lineBuffer = []` (line 64)
2. `_decoder` is also initialized as `null`: `this._decoder = null` (line 69)
3. The decoder is only created inside `_transform` when processing a buffer chunk (lines 90-92):
   ```typescript
   if (!this._decoder || this._decoderEncoding !== encoding) {
     this._decoder = new StringDecoder(encoding)
     this._decoderEncoding = encoding
   }
4. In the same `_transform` call, the chunk is split and concatenated to `_lineBuffer` (lines 100, 118):
   ```
   const lines = chunk.split(/\r\n|\r|\n/g)
   // ...
   this._lineBuffer = this._lineBuffer.concat(lines)
   ```
5. `String.prototype.split()` always returns an array with at least one element, even for empty strings:
    - `''.split(/\n/g)` → `['']`
    - `'text'.split(/\n/g)` → `['text']`
    - `'a\nb'.split(/\n/g)` → `['a', 'b']`

**Therefore:** If `this._decoder` exists (is not `null`), it means `_transform` was called at least once, which means `_lineBuffer` has at least one element from the `split()` operation. The condition `if (this._decoder)` (line 132) implicitly guarantees that `this._lineBuffer.length > 0`, making the explicit length check technically redundant but kept for defensive programming and code clarity. Edge case that cannot occur: The scenario where `_decoder` exists but `_lineBuffer` is empty would require the decoder to be created without going through `_transform`, which doesn't happen in this implementation. The current implementation correctly handles all cases.

aqrln commented:
Thanks so much for this! Could we also deduplicate this and put it in a single place while we're here? Even better, any chance we could use the built-in `node:readline` module instead?

onozaty committed:
refactor(client,internals,migrate,generator-helper): replace byline with node:readline

Replace custom byline implementation with Node.js built-in readline module.

- Remove 4 duplicate byline.ts implementations and tests
- Update BinaryEngine, SchemaEngineCLI, GeneratorProcess to use readline
- Remove byline exports from internals and migrate packages

onozaty committed:
chore: remove byline.ts from eslint ignore patterns

onozaty committed:
test(generator-helper): add readline multibyte UTF-8 handling tests

aqrln commented on packages/client/src/runtime/core/engines/binary/BinaryEngine.ts:
This file is gone on `main` so this is causing conflicts, please rebase

onozaty committed:
docs(generator-helper): improve test documentation for UTF-8 handling

- Remove PR-specific context from test comments
- Add reference to original issue #27695 for regression test clarity

aqrln reviewed (approved):
This is very nice, thank you for working on this and sorry for the delay!
